ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,664 @@
+        tensors of shape (n\ :sub:`cls`,) containing score or confidence of the predictions (doesn't need to be in
+        [0,1]). If there is no prediction or ground truth for a class, it could be absent from (TP, FP, scores) and P
+        dictionaries respectively.
+
+        Args:
+            pred: First member of :meth:`update`'s input is given as this argument. There's no constraint on its type
+                and shape.
+            target: Second member of :meth:`update`'s input is given as this argument. There's no constraint on its type
+                and shape.
+
+        Returns:
+            `(TP, FP, P, scores)` A quadrople of true positives, false positives, number of actual positives and scores.
+        """
+        raise NotImplementedError(
+            "Please subclass MeanAveragePrecision and implement `do_matching` method" " to use the metric in detection."
+        )
+
+    @reinit__is_reduced
+    def update(self, output: Union[Tuple[Any, Any], Tuple[torch.Tensor, torch.Tensor]]) -> None:
+        """Metric update function using prediction and target.
+
+        Args:
+            output: a binary tuple. It should consist of prediction and target tensors in the classification case but
+                for detection it is the same as the implemented-by-user :meth:`do_matching`.
+
+                For classification, this metric follows the same rules on ``output`` members shape as the
+                :meth:`Precision.update <precision.Precision.update>` except for ``y_pred`` of binary and multilabel
+                data which should be comprised of positive class probabilities here.
+        """
+
+        if self._task == "classification":
+            self._check_shape(output)
+            prediction, target = output[0].detach(), output[1].detach()
+            self._check_type((prediction, target))
+            scores, P = self._classification_prepare_output(prediction, target)
+            cast(List[torch.Tensor], self._scores).append(scores.to(self._device))
+            cast(List[torch.Tensor], self._P).append(
+                P.to(self._device, dtype=torch.uint8 if self._type != "multiclass" else torch.long)
+            )
+        else:
+            tps, fps, ps, scores_dict = self.do_matching(output[0], output[1])
+            self._check_matching_output_shape(tps, fps, scores_dict)
+            for cls in tps:
+                self._tp[cls].append(tps[cls].to(device=self._device, dtype=torch.uint8))
+                self._fp[cls].append(fps[cls].to(device=self._device, dtype=torch.uint8))
+                cast(Dict[int, List[torch.Tensor]], self._scores)[cls].append(scores_dict[cls].to(self._device))
+            for cls in ps:
+                cast(Dict[int, int], self._P)[cls] += ps[cls]
+            classes = tps.keys() | ps.keys()
+            if classes:
+                self._num_classes = max(max(classes) + 1, self._num_classes)
```

Comment:
Same comment here about task dispatching. ObjectionDetectionMAP is responsible for detection task so it can just override update method instead of this if/else

Reply:
I wondered if we add `DetectionMeanAveragePrecision` but the current unity seems more interesting.